### PR TITLE
Daily generation and upload of data warehouse export to S3

### PR DESCRIPTION
**Pull request by tekin**
This adds the code to upload a daily data warehouse export to S3.

The export is performed as a background job at 11pm, as requested by James R in the BI team.

The most effective way to review this pull request may be commit by commit as much of the context is included in the commit messages.

## Pre-Requisite

Before this can ship, the following environment variables need to be set on staging and production:

```bash
# production
AWS_S3_EXPORT_BUCKET=dss-infrastructure-production-data-warehouse
# staging
AWS_S3_EXPORT_BUCKET=dss-infrastructure-staging-data-warehouse
```

### CODE REVIEW

File: app/models/export/anything.rb [hunk 1]
```diff
@@ -44,25 +38,14 @@ def model_plural
...
-        STDERR.puts("Exporting #{model_plural} to #{filename}")
-        File.open(filename, 'w+') do |file|
-          yield file
-        end
-      when :stdout
-        yield STDOUT
-      else
-        STDERR.puts usage
+      STDERR.puts("Exporting #{model_plural} to #{filename}")
+      File.open(filename, 'w+') do |file|
```
Comment (james) [inline app/models/export/anything.rb]: Yay! Does this mean the specs will stop churning stuff out to the console too?

Comment (tekin) [reply app/models/export/anything.rb]: 👍

Comment (james) [reply app/models/export/anything.rb]: oh, you say it does in the commit message. 🎉

File: spec/models/data_warehouse_export_spec.rb [hunk 2]
```diff
@@ -33,7 +33,7 @@
     let!(:submission) { create(:completed_submission, framework: framework) }
     let!(:task) { submission.task }
 
-    before do
+    after do
```
Comment (james) [inline spec/models/data_warehouse_export_spec.rb]: 👍

File: app/models/export/relation.rb [hunk 3]
```diff
@@ -13,10 +13,10 @@ def run
         return
       end
 
-      log "Exporting #{model_plural} to #{filename}"
-      File.new(filename, 'w+').tap do |file|
+      log "Exporting #{model_plural}"
+      Tempfile.new([model_plural, '.csv']).tap do |file|
         export_class.new(@relation, file).run
-        file.close
+        file.rewind
```
Comment (james) [inline app/models/export/relation.rb]: Out of interest, what's the advantage of calling rewind here?

Comment (tekin) [reply app/models/export/relation.rb]: Without calling `rewind` first, calling `read` would return a blank string.

Comment (james) [reply app/models/export/relation.rb]: OK, that's definitely fair.

**Pull request merged**